### PR TITLE
#230 TestNg First Commit

### PR DESCRIPTION
**Pull request by cvadityaflipkart**
#230 TestNg First Commit

### CODE REVIEW

**Comment by tommysitu** [comment]
Epic! I'll review this soon-ish.

File: testng/src/main/resources/cert.pem [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+5u4cHwIDAQABo3IwcDAOBgNVHQ8BAf8EBAMCAqQwEwYDVR0lBAwwCgYIKwYBBQUH
+AwEwDwYDVR0TAQH/BAUwAwEB/zAdBgNVHQ4EFgQUxTebj9Kv16fuWngIO4zfjddv
+7fUwGQYDVR0RBBIwEIIOaG92ZXJmbHkucHJveHkwDQYJKoZIhvcNAQELBQADggEB
+AKOlihA/DIAc7soiPb8s5eLvY/YTqASgzy3S1oaqEsEFAnNPOu53ePNid6bmKDvD
+hc0E+sphPpcuWyzoSp4Nz5TFl7LTtIzU49mR+/Gn3tDucGLutS0PbdFen7swKTMO
+/HwXy+Bm2a9g4ewgJbfIf4MhgrdX6M4gJqMVL7q/NKeppHlQ4pBYFc+HjrF4V98V
+x/mHLc65qOh7iKPBVY7lSYipnMxRu5N7Q88eaqfaVh44xsIYO83N9Pn1uE9GOUu7
+Eb2Tc6UidvXZMWAfkkrGVyrJGWE4wTAIT/Dz1AKDPuTp16SyljaOZ2YahmFXMp3C
+Fj+GKkpM2WS40fUI9z40WGI=
+-----END CERTIFICATE-----
```
Comment (tommysitu) [inline testng/src/main/resources/cert.pem]: I don't think this is needed. There is one in the root project already

**Comment by tommysitu** [comment]
Can you git ignore these files:
```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   testng/src/test/resources/hoverfly/recorded-simulation.json

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        testng/src/test/resources/hoverfly/incremental-capture.json

```
git found these changes after running the tests

File: testng/src/main/java/io/specto/hoverfly/testng/HoverflyExtension.java [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+import java.util.*;
+import java.util.function.Predicate;
+
+import static io.specto.hoverfly.junit.core.HoverflyConfig.localConfigs;
+import static io.specto.hoverfly.junit.core.HoverflyConstants.DEFAULT_HOVERFLY_EXPORT_PATH;
+import static io.specto.hoverfly.junit.core.HoverflyMode.*;
+import static io.specto.hoverfly.junit.core.SimulationSource.file;
+import static io.specto.hoverfly.testng.HoverflyExtensionUtils.*;
+
+public class HoverflyExtension {
```
Comment (tommysitu) [inline testng/src/main/java/io/specto/hoverfly/testng/HoverflyExtension.java]: The first time I looked at it I thought it's a copy of the Junit 5 extension 😆  It'd be good to make the name more idiomatic for the TestNg world.

File: testng/src/test/java/io/specto/hoverfly/testng/HoverflyDslMatcherTest.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+import static io.specto.hoverfly.junit.dsl.ResponseCreators.*;
+import static io.specto.hoverfly.junit.dsl.matchers.HoverflyMatchers.*;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+import static org.springframework.http.HttpStatus.*;
+import static org.springframework.http.MediaType.APPLICATION_JSON;
+import static org.springframework.http.MediaType.APPLICATION_XML;
+
+@Listeners(HoverflyListener.class)
+public class HoverflyDslMatcherTest {
```
Comment (tommysitu) [inline testng/src/test/java/io/specto/hoverfly/testng/HoverflyDslMatcherTest.java]: I don't think you need to repeat these tests as it's for testing the DSL defined in the root project

**Comment by tommysitu** [comment]
You can exclude some tests from your subproject if they are not testing anything specific to the TestNg rules, like: 
HoverflyDslMatcherTest
HoverflyDslResponseTemplatingTest
HoverflyDslTest
HoverflyDslWithDelayTest
HoverflyDslWithStateTest
....

Some unit tests on your `HoverflyExtension` APIs and make sure they configured hoverfly properly would be enough. Otherwise, there will be lots of duplicated tests to maintain.

File: testng/src/main/java/io/specto/hoverfly/testng/HoverflyExtension.java (hunk 2 above)
Comment (cvadityaflipkart) [reply testng/src/main/java/io/specto/hoverfly/testng/HoverflyExtension.java]: Changed to HoverflyTestNG

**Comment by cvadityaflipkart** [comment]
> Thank you for such an epic contribution! I'm not a TestNg user myself, but the implementations look neat and more importantly it works! You'll need to help supporting it please 🙏
> 
> Can you also provide a README? You can create one based on https://github.com/SpectoLabs/hoverfly-java/blob/master/junit5/README.adoc
> 
> Meanwhile I'll sort out the release and will look at how to reduce duplicated code, extracting some interfaces or reusable classes from the root project.
> 
> Great work! 🌟

Happy to contribute. Will create the README on this branch. Is it fine ?

**Comment by tommysitu** [comment]
you may need to do a rebase, but give it a go and raise a new PR.

**Pull request merged**